### PR TITLE
ItemDictionary<T>.Enumerator has an unnecessary finalizer

### LINKED ISSUE #7208: ItemDictionary<T>.Enumerator has an unnecessary finalizer
This finalizer increases the cost of the object, but does not provide any value:

https://github.com/dotnet/msbuild/blob/b827bf58c21f7a38770d786848c76f771d995a94/src/Build/Collections/ItemDictionary.cs#L408-L414

Both the finalizer and the `Dispose(bool)` method should be eliminated from this type.

### PR DESCRIPTION
**Pull request by elachlan**
Fixes #7208

### CODE REVIEW

File: src/Build/Collections/ItemDictionary.cs [hunk 1]
```diff
@@ -464,31 +456,9 @@ public void Reset()
...
-            /// </summary>
-            private void Dispose(bool disposing)
-            {
-                if (disposing)
-                {
-                    if (_listEnumerator != null)
-                    {
-                        if (_itemEnumerator != null)
-                        {
-                            _itemEnumerator.Dispose();
```
Comment (rainersigwald) [inline src/Build/Collections/ItemDictionary.cs:482]: Don't we still need to dispose these owned resources?

Comment (elachlan) [reply src/Build/Collections/ItemDictionary.cs:482]: I wasn't sure, so I have added it back.

File: src/Build/Collections/ItemDictionary.cs [hunk 2]
```diff
@@ -466,29 +458,18 @@ public void Reset()
...
-                        _listEnumerator.Dispose();
-                        _listEnumerator = null;
+                        _itemEnumerator.Dispose();
+                        _itemEnumerator = null;
                     }
+
+                    _listEnumerator.Dispose();
+                    _listEnumerator = null;
                 }
+                GC.SuppressFinalize(this);
```
Comment (ladipro) [inline src/Build/Collections/ItemDictionary.cs:472]: This has no effect when the class doesn't have a finalizer, please remove.

**Review by ladipro — CHANGES_REQUESTED** [review]

**Review by ladipro — APPROVED** [review]

**Comment by KalleOlaviNiemitalo** [comment]
This looks like the nested enumerator type could be replaced with an iterator method: nested `foreach` loops and `yield return`. I don't know how the performance would be, though.

Iterators do not support IEnumerator.Reset() but the current implementation of Reset is bogus anyway: it resets _itemEnumerator too, so if you had the hierarchy ((A B C) (D E F)) and the current item is E, then after the reset, it will enumerate (D E F A B C D E F) even though the right answer would be just (A B C D E F).

**Comment by ladipro** [comment]
@KalleOlaviNiemitalo that's indeed a bug, thank you. It looks like we should do

```C#
_itemEnumerator = GetNextItemEnumerator();
```

instead of resetting whatever `_itemEnumerator` is currently pointing to. Out of scope of this PR but would be nice to fix for sure. Also +1 on trying to replace the class with an iterator.

**Comment by KalleOlaviNiemitalo** [comment]
In principle, Reset and MoveNext should also call `_itemEnumerator?.Dispose()` before assigning `_itemEnumerator = GetNextItemEnumerator()`. In practice, _itemEnumerator is always a LinkedList\<T>.Enumerator whose Dispose() method would do nothing anyway. Although LinkedList\<T> is not sealed, ItemDictionary\<T> does not let its callers add instances of any derived types to _itemLists.

**Comment by elachlan** [comment]
Could one of you please write up an issue to spec out this suggestion? It sounds like a good idea.

**Comment by KalleOlaviNiemitalo** [comment]
Filed <https://github.com/dotnet/msbuild/issues/7286>.

**Pull request merged**